src/timely-util/src/builder_async.rs

```diff
@@ -0,0 +1,361 @@
+    input_handles: Rc<Cell<usize>>,
+}
+
+/// An async Waker that activates a specific operator when woken and marks the task as ready
+struct TimelyWaker {
+    activator: SyncActivator,
+    active: AtomicBool,
+    task_ready: AtomicBool,
+}
+
+impl ArcWake for TimelyWaker {
+    fn wake_by_ref(arc_self: &Arc<Self>) {
+        // Only activate the timely operator if it's not already active to avoid an infinite loop
+        if !arc_self.active.load(Ordering::SeqCst) {
+            arc_self.activator.activate().unwrap();
+        }
+        arc_self.task_ready.store(true, Ordering::SeqCst);
+    }
+}
+
+/// Async handle to an operator's input stream
+pub struct AsyncInputHandle<T: Timestamp, D: Container, P: Pull<BundleCore<T, D>>> {
+    /// The underying synchronous input handle
+    sync_handle: InputHandleCore<T, D, P>,
+    /// Frontier information of input streams shared with the operator. Each frontier is paired
+    /// with a flag indicating whether or not the handle has seen the updated frontier.
+    shared_frontiers: Rc<RefCell<Vec<(Antichain<T>, bool)>>>,
+    /// The index of this handle into the shared frontiers vector
+    index: usize,
+    /// Reference to the reactor queue of this input handle where Wakers can be registered
+    reactor_registry: Rc<RefCell<Vec<Waker>>>,
+    /// Reference to the count of active read handles
+    input_handles: Rc<Cell<usize>>,
+}
+
+impl<T: Timestamp, D: Container, P: Pull<BundleCore<T, D>>> Drop for AsyncInputHandle<T, D, P> {
+    fn drop(&mut self) {
+        self.input_handles.set(self.input_handles.get() - 1);
+    }
+}
+
+impl<T: Timestamp, D: Container, P: Pull<BundleCore<T, D>>> AsyncInputHandle<T, D, P> {
+    /// Produces a future that will resolve to the next event of this input stream
+    ///
+    /// # Cancel safety
+    ///
+    /// This method is cancel-safe
+    pub fn next(&mut self) -> impl Future<Output = Option<Event<'_, T, D>>> + '_ {
+        NextFut { handle: Some(self) }
+    }
+}
+
+/// The future returned by `AsyncInputHandle::next`
+struct NextFut<'a, T: Timestamp, D: Container, P: Pull<BundleCore<T, D>>> {
+    handle: Option<&'a mut AsyncInputHandle<T, D, P>>,
+}
+
+/// An event of an input stream
+pub enum Event<'a, T: Timestamp, D> {
+    /// A data event
+    Data(CapabilityRef<'a, T>, RefOrMut<'a, D>),
+    /// A progress event
+    Progress(Antichain<T>),
+}
+
+impl<'a, T: Timestamp, D: Container, P: Pull<BundleCore<T, D>>> Future for NextFut<'a, T, D, P> {
+    type Output = Option<Event<'a, T, D>>;
+
+    fn poll(mut self: Pin<&mut Self>, cx: &mut Context) -> Poll<Self::Output> {
+        let handle = self.handle.take().expect("future polled after completion");
+        match handle.sync_handle.next() {
+            Some((cap, data)) => {
+                // SAFETY: This is necessary only due to a limitation in the borrow checker. Once
+                // Rust starts using the polonius borrow checker the transmute can be removed.
```

Comment:
I'd link to the tokio code that does the same thing, with the explanation

I still need to convince myself this is correct

Reply:
Is this correct? the lifetime of the `cap` and `data` are tied to the `next` call right? which goes all the way up to the `self` borrow in `Pull::pull`?

Reply:
No, the lifetime of `cap` and `data` are in fact `'a` and not `'self`. This is because this:

```rust
let handle = self.handle.take().expect(..);
```

calls the function `fn for<'self> take(self: &'self mut Self) -> T` on `Option` and for us `T == &'a mut AsyncInputHandle<T, D, P>`. Therefore by putting a `None` there we get a `'a` lifetimed handle on our stack.

Then, using that `'a` lifetimed handle on the stack we call `InputHandle::next(&'a mut handle) -> (CapabilityRef<'a>, RefOrMut<'a, _>)` and we get back `'a` lifetimed capabilities and data. Rust fails to do this inference correctly but that's why it's safe to transmute them.